File: Tests/AINotebookCoreTests/NotebookStoreEmbeddingsTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class NotebookStoreEmbeddingsTests: XCTestCase {

    func testStoreAndLoadEmbedding() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "x", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [
                ChunkDraft(text: "alpha", tokenCount: 1),
                ChunkDraft(text: "beta",  tokenCount: 1)
            ]
        )
        let chunks = try store.chunks(sourceId: s.id!)
        let v1 = EmbeddingVector(values: [1, 0, 0, 0])
        let v2 = EmbeddingVector(values: [0, 1, 0, 0])
        try store.storeEmbedding(chunkId: chunks[0].id!, model: "m", vector: v1)
        try store.storeEmbedding(chunkId: chunks[1].id!, model: "m", vector: v2)

        let loaded = try store.embeddings(notebookId: nb.id!, model: "m")
        XCTAssertEqual(loaded.count, 2)
        let pairs = Dictionary(uniqueKeysWithValues: loaded.map { ($0.chunkId, $0.vector.values) })
        XCTAssertEqual(pairs[chunks[0].id!], v1.values)
        XCTAssertEqual(pairs[chunks[1].id!], v2.values)
    }

    func testUnembeddedChunksReturnsOnlyMissingForModel() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "x", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: (0..<3).map { ChunkDraft(text: "c\($0)", tokenCount: 1) }
        )
        let chunks = try store.chunks(sourceId: s.id!)
        try store.storeEmbedding(
            chunkId: chunks[0].id!,
            model: "m",
            vector: EmbeddingVector(values: [0])
        )
        let pending = try store.unembeddedChunks(model: "m", limit: 100)
        let ids = Set(pending.map(\.id))
        XCTAssertEqual(ids, Set([chunks[1].id!, chunks[2].id!]))
    }

    func testReplaceEmbeddingOverwrites() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "x", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [ChunkDraft(text: "a", tokenCount: 1)]
        )
        let chunk = try store.chunks(sourceId: s.id!).first!
        try store.storeEmbedding(
            chunkId: chunk.id!,
            model: "m",
            vector: EmbeddingVector(values: [1, 0])
        )
        try store.storeEmbedding(
            chunkId: chunk.id!,
            model: "m",
            vector: EmbeddingVector(values: [0, 1])
        )
        let loaded = try store.embeddings(notebookId: nb.id!, model: "m")
        XCTAssertEqual(loaded.count, 1)
        XCTAssertEqual(loaded[0].vector.values, [0, 1])
    }

    func testDeleteAllEmbeddingsForModelClearsOnlyThatModel() throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "x", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [ChunkDraft(text: "a", tokenCount: 1)]
        )
        let chunk = try store.chunks(sourceId: s.id!).first!
        try store.storeEmbedding(
            chunkId: chunk.id!, model: "m1", vector: EmbeddingVector(values: [1])
        )
        try store.deleteAllEmbeddings(model: "m1")
        XCTAssertEqual(try store.embeddings(notebookId: nb.id!, model: "m1").count, 0)
    }
}

File: windows/tests/AINotebook.Core.Tests/Storage/NotebookStoreEmbeddingsTests.cs
using AINotebook.Core.Models;
using AINotebook.Core.Storage;
using Xunit;

namespace AINotebook.Core.Tests.Storage;

public class NotebookStoreEmbeddingsTests
{
    private static (NotebookStore store, long nb, long src) Seed(int chunkCount)
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N").Id!.Value;
        var src = store.CreateSource(nb, SourceType.Text, "t", null, null).Id!.Value;
        var drafts = Enumerable.Range(0, chunkCount)
            .Select(i => new ChunkDraft($"chunk {i}", 1, null)).ToList();
        store.ReplaceChunks(src, drafts);
        return (store, nb, src);
    }

    [Fact]
    public void StoreAndLoadEmbedding()
    {
        var (store, nb, src) = Seed(2);
        using (store)
        {
            var chunks = store.Chunks(src);
            store.StoreEmbedding(chunks[0].Id!.Value, "m", new EmbeddingVector(new[] { 0.1f, -0.2f }));
            store.StoreEmbedding(chunks[1].Id!.Value, "m", new EmbeddingVector(new[] { 3.14f, -42.0f }));
            var loaded = store.Embeddings(nb, "m").OrderBy(e => e.ChunkId).ToList();
            Assert.Equal(2, loaded.Count);
            Assert.Equal(src, loaded[0].SourceId);
            Assert.Equal(new[] { 0.1f, -0.2f }, loaded[0].Vector.Values);
            Assert.Equal(new[] { 3.14f, -42.0f }, loaded[1].Vector.Values);
        }
    }

    [Fact]
    public void UnembeddedChunksReturnsOnlyMissingForModel()
    {
        var (store, _, src) = Seed(3);
        using (store)
        {
            var chunks = store.Chunks(src);
            store.StoreEmbedding(chunks[0].Id!.Value, "m", new EmbeddingVector(new[] { 1f }));
            var unembedded = store.UnembeddedChunks("m", 100);
            Assert.Equal(2, unembedded.Count);
            Assert.Equal(new[] { chunks[1].Id!.Value, chunks[2].Id!.Value },
                unembedded.Select(c => c.Id!.Value).ToArray());
            Assert.Equal(2, store.UnembeddedCount("m"));
        }
    }

    [Fact]
    public void ReplaceEmbeddingOverwrites()
    {
        var (store, nb, src) = Seed(1);
        using (store)
        {
            var chunkId = store.Chunks(src)[0].Id!.Value;
            store.StoreEmbedding(chunkId, "m", new EmbeddingVector(new[] { 9f, 9f }));
            store.StoreEmbedding(chunkId, "m", new EmbeddingVector(new[] { 0f, 1f }));
            var loaded = store.Embeddings(nb, "m");
            Assert.Single(loaded);
            Assert.Equal(new[] { 0f, 1f }, loaded[0].Vector.Values);
        }
    }

    [Fact]
    public void DeleteAllEmbeddingsForModelClearsOnlyThatModel()
    {
        var (store, nb, src) = Seed(1);
        using (store)
        {
            var chunkId = store.Chunks(src)[0].Id!.Value;
            store.StoreEmbedding(chunkId, "m1", new EmbeddingVector(new[] { 1f }));
            store.DeleteAllEmbeddings("m1");
            Assert.Empty(store.Embeddings(nb, "m1"));
        }
    }
}
